node/src/components/consensus/protocols/highway/round_success_meter.rs

```diff
@@ -0,0 +1,146 @@
+    rounds: VecDeque<bool>,
+    current_round_id: u64,
+    proposals: Vec<C::Hash>,
+    min_round_exp: u8,
+    current_round_exp: u8,
+    current_round_length: u64,
+}
+
+impl<C: Context> RoundSuccessMeter<C> {
+    pub fn new(round_exp: u8, min_round_exp: u8, timestamp: Timestamp) -> Self {
+        let current_round_length = 1u64 << round_exp;
+        let current_round_index = timestamp.millis() / current_round_length;
+        let current_round_id = current_round_index * current_round_length;
+        Self {
+            rounds: VecDeque::with_capacity(NUM_ROUNDS_TO_CONSIDER),
+            current_round_id,
+            proposals: Vec::new(),
+            min_round_exp,
+            current_round_exp: round_exp,
+            current_round_length,
+        }
+    }
+
+    fn change_exponent(&mut self, new_exp: u8, timestamp: Timestamp) {
+        self.rounds = VecDeque::with_capacity(NUM_ROUNDS_TO_CONSIDER);
+        self.current_round_exp = new_exp;
+        self.current_round_length = 1u64 << new_exp;
+        let round_index = timestamp.millis() / self.current_round_length;
+        self.current_round_id = round_index * self.current_round_length;
+        self.proposals = Vec::new();
+    }
+
+    fn check_proposals_success(&self, state: &State<C>, proposal_h: &C::Hash) -> bool {
+        let total_w = state.total_weight();
+
+        let finality_detector =
+            FinalityDetector::<C>::new(max(total_w / 100 * THRESHOLD, Weight(1)));
+
+        // check for the existence of a level-1 summit
+        finality_detector.find_summit(1, proposal_h, state) == 1
+    }
+
+    /// Registers a proposal within this round - if it's finalized within the round, the round will
+    /// be successful.
+    pub fn new_proposal(&mut self, proposal_h: C::Hash, timestamp: Timestamp) {
+        // only add proposals from within the current round
+        if timestamp.millis() >= self.current_round_id
+            && timestamp.millis() - self.current_round_id < self.current_round_length
+        {
+            self.proposals.push(proposal_h);
+        }
+    }
+
+    /// If the current timestamp indicates that the round has ended, checks the known proposals for
+    /// a level-1 summit.
+    /// If there is a summit, the round is considered successful. Otherwise, it is considered
+    /// failed.
+    /// Next, a number of last rounds are being checked for success and if not enough of them are
+    /// successful, we return a higher round exponent for the future.
+    /// If the exponent shouldn't grow, and the round ID is divisible by a certain number, a lower
+    /// round exponent is returned.
+    pub fn calculate_new_exponent(&mut self, state: &State<C>) -> u8 {
+        let now = Timestamp::now();
+        // if the round hasn't finished, just return whatever we have now
+        if now.millis().saturating_sub(self.current_round_id) < self.current_round_length {
+            return self.new_exponent();
+        }
```

Comment:
Can't we make it so that the caller knows when to check for new round exponent, rather than deciding it based on `Timestamp::now()` inside the method?

Reply:
From our point of view, a round might have finished earlier, and I think it's important to take that into account. For example, we could get the final message necessary for finalization before we get the next proposal, but after the round has already finished according to our clock. In such a case, I think we should still be counting the round as failed, as it indicates that we (the network) were too slow.